demos/https/iot_demo_https_s3_download_async.c

```diff
@@ -0,0 +1,741 @@
+        /* We may get an error where the server closed the connection, but data was still filled
+           in the buffer. So we do not want to necessarily cancel the request. */
+    }
+
+    /* Process the body here. */
+    rangeValueStr = pDownloadData->rangeValueStr;
+    IotLogInfo( "Response return code: %d for %s", status, rangeValueStr );
+    IotLogInfo( "Response Body for %s:\r\n%.*s", rangeValueStr, readLen, _pRespBodyBuffer );
+
+    IotHttpsClient_ReadContentLength( respHandle, &contentLength );
+    if( contentLength == 0)
+    {
+        IotLogError("Failed to retrieve the increment %s. The content length was %d.", rangeValueStr, contentLength);
+    }
+
+    /* If the content length of the message is not equal to the size of the byte range we want to download then 
+       cancel the request. */
+    if (contentLength != pDownloadData->numReqBytes)
+    {
+        IotLogError("The Content-Length found in this file does not equal the number of bytes requested. So we may not \
+            download the file completely. The content length is %d and the requested number of bytes for this request \
+            is %d", contentLength, pDownloadData->numReqBytes);
+        IotHttpsClient_CancelResponseAsync(respHandle);
+    }
+
+    /* This callback could be invoked again if there is still more data on the network to be read for this response, so 
+       we increment the current amount downloaded. */
+    pDownloadData->currDownloaded += readLen;
+}
+
+/*-----------------------------------------------------------*/   
+
+/**
+ * @brief Callback for an asynchronous request to notify that the response is complete.
+ * 
+ * @param[in] pPrivData - User private data configured with the HTTPS Client library request configuration.
+ * @param[in] respHandle - Identifier for the current response finished.
+ * @param[in] rc - Return code from the HTTPS Client Library signalling a possible error.
+ * @param[in] status - The HTTP response status.
+ */
+static void _responseCompleteCallback( void * pPrivData, IotHttpsResponseHandle_t respHandle, IotHttpsReturnCode_t rc, uint16_t status )
+{
+    _asyncDownloadData_t* pDownloadData = ( _asyncDownloadData_t* )( pPrivData );
+
+    ( void )respHandle;
+    ( void )rc;
+    ( void )status;
+
+    IotLogInfo( "Part %s has been fully processed.", pDownloadData->rangeValueStr );
+    
+    /* Free up this request from the request pool. */
+    _pInUseRequests[pDownloadData->reqNum] = false;
+
+    /* If in this response the total amount read does not equal the number of bytes we requested, then something went 
+       wrong. */
+    if( pDownloadData->currDownloaded != pDownloadData->numReqBytes )
+    {
+        IotLogError( "There was a problem downloading the file. We downloaded %d. but wanted %d.", 
+        pDownloadData->currDownloaded,
+        pDownloadData->numReqBytes );
+        IotSemaphore_Post( &( _fileFinishedSem ) );
```

Comment:
Should the code exit here go to common cleanup where semaphore post is called?   It is possible that (_bytesFileDownloadedSoFar >= _fileSize) is true (line 451)  as well as  (pDownloadData->currDownloaded != pDownloadData->numReqBytes ) is true, in that case,  IotSemaphore_Post() will be called two times.

Reply:
It is not a problem if IotSemaphorePost() is called twice but, I will put the other part in a else case because it shoudl not be run if there was an error. Thanks.